### ISSUE TITLE
Can you decrypt caption file to srt subtitle

### ISSUE THREAD

**Issue by kimvuchannel** [h4ck-rOOt/Lynda-Decryptor#9]
Please help me. I think you can do it :)

**Comment by dungtd91** [comment]
Please help me. Decrypt caption file to srt

**Comment by mdomnita** [comment]
Maybe I will try to do it if nobody started to, shouldn't be very difficult. Studied the file structure a bit and does not seem to be a big issue. Just got to get used to the project a bit as I am not a VS developer.

**Comment by h4ck-rOOt** [comment]
@dungtd91 @kimvuchannel please take a look at this little app [Lynsub](http://www.softpedia.com/get/Multimedia/Video/Other-VIDEO-Tools/Lynsub.shtml).
Maybe this will help you out. Personally i don't know what files needed to decrypt them to srt format.

So it's necessary to take a look at the original source code of the lynda.com app... but i have spare time left atm, so it would be nice if @mdomnita can solve this.

**Comment by kimvuchannel** [comment]
@h4ck-rOOt A few months ago, lynda subtitle creation is easy...but lynda have fixed. Free user can't view transcript of many course or view with text format. Lyndasub.ir is down. I have download lynsub app but it not working :(

**Comment by mdomnita** [comment]
@h4ck-rOOt , this is the main issue with the subtitles. They are there in the .caption files in the folder. I don't seem to be able to find a way to link the .caption file to the .lynda file. There is nothing in the DB and I did not notice any kind of connection. The only way I could do it is by ordering the files by date created (both the .lynda and the .caption and creating a Hash (dictionary, list, whatever) that links the .lynda file to the .caption. This is the biggest issue and I don;t know if this approach is correct, it's actually a workaround that presumes the .caption files were downloaded in the same order as the videos.
Other than this, the timestamps are in the .caption, the text is in the .caption. Seems to me everything that has to be done is cleaning the .caption file, removing the "useless" characters, creating a list with timestamps and generating the .srt file according to the pretty clear and simple .srt specific format.
As I said, I will try to do it. Those "useless" characters appear different in different viewers, I have no idea on what info is there and what the Lynda app does with it. However, all the subtitle text and timestamps are in plain text.

**Comment by mdomnita** [comment]
Well, I did it. I mean, sort of. I wrote the code to convert from .caption files to .srt files, I was not able to figure out the logic behind how the desktop app links the .lynda files to the .srt files. Also, not all the characters that remained there from the binary structure of the .caption file were removed. But it works and it's readable.
Get and try the thing from here https://goo.gl/KyE8yA
Just do LyndaCaptionToSrtConvertor.exe /D your_dir_with_caption_files and you will get the subtitles in the same dir. 
And the repository is here. https://github.com/mdomnita/LyndaCaptionToSrtConvertor with everything explained in the comments.
It's a very crude solution but it works and maybe someone who has C# skills can help me improve it. As I said, I usually develop Perl and JS but I did this in C# so that someone may probably improve it a bit and  include it in this project.

**Comment by ktsarlz** [comment]
@mdomnita Appreciate it man! There are some extra letters, But they are readable. The only trouble I come across with is that the captions files does not have same filename to the .lynda files, hence testing each and everyone is kinda tedious Haha

**Comment by mdomnita** [comment]
Yeah, I know that. i will ry to find a way to relate the .lynda files to the .caption but I did not find out how it works yet. Ițll keep you notified if I do

**Comment by h4ck-rOOt** [comment]
It would ne nice to have some caption files to do this kind of magic and translate them depending on database entries.

Could someone provide these files, tests or even compare encrypted and decrypted files? There must be a hint in their code (try reverse engineering).

**Comment by mdomnita** [comment]
I further cleaned the file and removed more or less all that is not actual subtitle text. The Program.cs is in the repo and every "move" is commented there.
Related to link between .caption files and .lynda files, a friend gave me a hint:
 protected string GetCaptionFileName(Video video)
        {
            return (Hash.MD5(video.ID.ToString()) + ".caption");
        }

**Comment by emon2196** [comment]
@mdomnita thanks u, i converted all sub file successful but I wonder how to know which sub is right for a video ? plz help <3

**Comment by Dev-iL** [comment]
@emon2196 the subtitle file name is the md5 hash of the video ID.

**Comment by emon2196** [comment]
@Dev-iL Thanks for reply but I don't get it. 
U mean sub file name = MD5 hash of video ?
I checked with course: " Illustrator CC 2015 one vs one advance" but it not correct

**Comment by Dev-iL** [comment]
@emon2196 This is absolutely not what I meant. Assuming the videoID is `513444` (you find this in the SQLite `.db` file), then the MD5 of this number is `e6c00103ac221a34d8e78c93ece94daa` and so the subtitle filename is that hash + `.caption`.
I hope it's clear now.

**Comment by mdomnita** [comment]
I managed to get the subtitle filename. It wasn't just a line of code like above but I got it right, anyway. I did  the conversion for a single file and it works, will work on full directory conversion tomorrow. I still need to improve/rewrite the srt convertor I did, after doing some reverse engineering on the application code. For the moment I get most of the subtitles right, about 90% but it's still something created based on trial and error, not on study and logic (the conversion from .caption to .srt format).
I will keep you updated if I manage to get some results and commit anything.

**Comment by mdomnita** [comment]
Ok, I did the file renaming and subtitle conversion. 
Added my own class that does subtitle conversion and created a pull request. 
Subtitle conversion (the actual .cption to .srt file conversion) still needs work but I will do it in the near future.
Check out the binary from my gDrive https://goo.gl/g4sAh5

**Comment by emon2196** [comment]
@mdomnita I tested your version. All videos have right subtiles <3 
But your version can't choose output folder, it decrypt direct to course folders. 
And the second is it didn't change name and put videos file into each folder like h4ck-rOOt's version.

Anyway, thank so much much much much ....<3

**Comment by mdomnita** [comment]
@emon2196 , for me it works. Just add the parameter /OUT <your output folder> when running from the console. 
Get the files from the link, unpack them somewhere.
This is how my directory with the LyndaDecryptor application looks like:
[http://i64.tinypic.com/2vv1h5x.jpg](url)

Run command like this (approx):
 .\LyndaDecryptor.exe **/DB** c:\Users\<your user>\AppData\Local\lynda.com\Lynda.com Desktop App\db
.sqlite **/D** c:\<wherever your courses are>\<course number>\ **/OUT** c:\...\<folder where you want the result>

This is about it. Just tested and it works. You need to specify at least /DB the right path for the sqlite.db file and the /D directory with your course, in which case the app will create the structure into a subdirectory named "decrypted" there (thanks @h4ck-rOOt for that)

**Comment by emon2196** [comment]
@mdomnita [http://imgur.com/a/iT7tQ](url)
Here my commend. My file decrypted did not go to E:lynda but go to courses folder

**Comment by mdomnita** [comment]
I think that the DB.sqlite path might not be good? DOes the same for me when I don't give a right path to the DB. Check it out in the command prompt, the command is something like this :

if exist "c:\Users\<youruser>\AppData\Local\lynda.com\Lynda.com Desktop App\db.sqlite" echo OK, it exists
 
Otherwise maybe it cannot find the course in your database? (can you see it in the Lynda desktop app and confirm everything's ok?). Will test further if everything looks ok from your side and try to figure out where your problem might be

**Comment by emon2196** [comment]
@mdomnita yah i think that the problem (Y) I deleted all my courses and db.sqlite then download again. Everything working like a charm now <3 Thank so much <3

**Comment by JPGrefaldo** [comment]
Thank you, for adding the sub.

**Comment by dhiaelhak47** [comment]
thanks :dagger:
